docs/design.md

```diff
@@ -272,89 +278,63 @@ Every minute, run the following high-level loop:
 1. Fetch the list of providers and their latest advertisements (heads) from
    https://cid.contact/providers. (This is **one** HTTP request.)
 
-2. Fetch the state of all providers from our database. (This is **one** SQL
-   query.)
-
-3. Update the state of each provider as described below, using the name
-   `LastAdvertisement` for the CID of the latest advertisement provided in the
-   response from cid.contact. (This is a small bit of computation plus **one**
-   SQL query.)
+2. Update the in-memory info we keep for each provider (address, CID of the last
+   advertisement).
 
 > **Note:** Instead of running the loop every minute, we can introduce a
 > one-minute delay between the iterations instead. It should not matter too much
-> in practice, though. I expect each iteration to finish within one minute:
->
-> - The three steps outlined above require only three interactions over
->   HTTP/SQL.
-> - The long chain walks are executed in the background and don't block this
->   loop.
+> in practice, though. I expect each iteration to finish within one minute, as
+> it's just a single HTTP call to cid.contact.
 
-For each provider listed in the response:
+**Walk advertisement chains (in background)**
 
-1. If `last_head` is not set, then we need to start the ingestion from scratch.
-   Update the state as follows and start the chain walker:
+The chain-walking algorithm runs in the background and loops over the following
+steps.
```

Comment:
The current implementation is a bit inefficient when it comes to Redis usage. I deployed it to Fly.io + Redis@Upstash and after 10 minutes of running, we made 212k requests costing us $0.42. That's $144/day or $4.3k/month.

<img width="179" alt="Screenshot 2024-09-16 at 16 17 36" src="https://github.com/user-attachments/assets/daf1d6f9-d814-4389-83a9-1460d68531a3">

I can avoid most of the `GET` requests by sharing the walker state between ` walkOneStep` invocations, but that's only a 33% saving ($1.4k, new monthly spend is $2.9k).

The next possible improvement is debouncing the writing of walker state updates, e.g., save the state only every 10 steps. That should save us another 90% of 1/3 commands ($1.3k, new monthly spend is $1.6k).

Spending $1.6k per month still feels too much to me.

What other options we have available:

- Fixed plan (100 reqs/s, 3GB storage) - I think this may be enough for now, until we learn how much data we need to store: $50 + ($10 x read region)

- Pro 2K (2000 reqs/s, 100GB storage) - I think this is a safe bet, hopefully we won't need more than 100GB: $280 + ($100 x read regions)

- Find a different storage solution, e.g. Postgres DB we use elsewhere. 

- Re-architect piece-indexer into a single service that's handling both walking advertisements and serving index requests from Spark checkers. Then we can persist the data in a mounted volume using a lightweight key/value store, e.g. IndexDB.

<img width="292" alt="Screenshot 2024-09-16 at 16 29 38" src="https://github.com/user-attachments/assets/429674e8-e9ba-4c76-8034-99e704fb5bc7">

Reply:
I think we can significantly reduce the number of `SADD` calls as well:

- Rework the schema from key=`piece-payload:${providerId}:${pieceCid}` value=set of payload CIDs to key=`piece-payload:${providerId}` value=hash mapping `${pieceCid}` to a single payload CID.
- Add new index entries in batches, e.g. every 10 steps.

Reply:
The remaining question is whether this will be good enough or whether we need to find something cheaper/more efficient than a managed Redis instance.

The downside of coupling the process ingesting advertisements with the process serving HTTP requests is that a spike in the incoming HTTP traffic can negatively affect the ingestion of new advertisements. Plus, it may be more difficult to scale this monolith, as we cannot add many cheap VM instances to handle the HTTP traffic - we will need to scale up a single VM.

Reply:
- I agree it's important to keep the processes decoupled, as otherwise decoupling later on will be painful.
- 4.3k$/m is a lot, it would be more expensive than the rest of our infrastructure combined
- 1.6k$/m is still a lot, since that was ~ the Influx price we decided we had to reduce
- I'm +1 on using prepaid plans: It lets us keep a simple design while having manageable spending
- I agree that a different data store could be cheaper, but I think the Redis design should be the easiest to manage (Redis is simple + we can use a managed instance).
- If switching to Postgres will make the overall system simpler and easier to manage, then this is interesting - our time is important and we want the solution that strikes a balance between cost and dev time. I'd be more happy to pay 1k$/m where we need to spend 0h maintaining it, than 100$/m where we're spending 8h/month on it.

Reply:
I hope we can find a solution that is simple and cheap: I would like not to introduce batching or other performance optimizations unless we know we really need them. They all make the system harder to debug.

Reply:
I switched Upstash to a paid plan. I also added 455b1ae to reduce the number of `GET` calls, it was an easy change that does not introduce too much complexity, IMO.

@juliangruber could you please review the code now?